### DISCUSSION TITLE
CR #1: Comment from Fabien Lavocat

### DISCUSSION THREAD

**Discussion by FabienLavocat** [ossrs/ffmpeg-webrtc#15] [General]
Review for https://github.com/ossrs/ffmpeg-webrtc/pull/1

I see that your readme section for [Millicast](https://github.com/ossrs/ffmpeg-webrtc/pull/1#usage-ffmpeg-milicast) is not ready yet, I was testing it and found out that the current version of the code is not working. It looks like it is missing the [PRIORITY](https://datatracker.ietf.org/doc/html/rfc8445#section-7.1.1) attribute in the STUN Request Binding. This attribute is mandatory (according to the RFC) and if missing, Millicast is not accepting the request.

To fix this, I've added the `STUN_ATTR_PRIORITY` attribute to the `STUNAttr` enum.
```cpp
/* STUN Attribute, comprehension-required range (0x0000-0x7FFF) */
enum STUNAttr {
    STUN_ATTR_USERNAME                  = 0x0006, /// shared secret response/bind request
    STUN_ATTR_PRIORITY                  = 0x0024, /// priority
    STUN_ATTR_USE_CANDIDATE             = 0x0025, /// bind request
    STUN_ATTR_MESSAGE_INTEGRITY         = 0x0008, /// bind request/response
    STUN_ATTR_FINGERPRINT               = 0x8028, /// rfc5389
};
```

Then, in the `ice_create_request` function, I add this attribute to the request, just after writing the `STUN_ATTR_USE_CANDIDATE` attribute.
```cpp
avio_wb16(pb, STUN_ATTR_PRIORITY); /* attribute type priority */
avio_wb16(pb, 4); /* size of priority */
ffio_fill(pb, av_lfg_get(&whip->rnd), 4); // Random value for now
```

The value is random for now, just because I did not have time yet to do the actual compute of that value following the [recommended formula](https://datatracker.ietf.org/doc/html/rfc8445#section-5.1.2.1).

**Comment by FabienLavocat** [comment]
Also, sending the [WebRTC heartbeat](https://github.com/ossrs/ffmpeg-webrtc/pull/1/files#diff-51c05c09836773348ba6c61e1b41f58659f7680870264110eec516fb2ba745cfR1658) is mandatory for streaming to Millicast. Simply call the following code every few seconds.
```cpp
ret = ice_create_request(s, whip->buf, sizeof(whip->buf), &size);
// ...
ret = ffurl_write(whip->udp_uc, whip->buf, size);
```

**Comment by winlinvip** [comment]
Thanks for you test. There is a new patch #18 . If I may, could you help to test it?
